src/lr_apply.rs

```diff
@@ -41,162 +41,152 @@ unsafe fn lr_stripe<BD: BitDepth>(
+        filter[1][3] = 128 - (filter[1][0] + filter[1][1] + filter[1][2]) * 2;
+
+        lr_fn = dsp.lr.wiener[((filter[0][0] | filter[1][0]) == 0) as usize];
     } else {
-        if !(lr.r#type as c_int == RAV1D_RESTORATION_SGRPROJ as c_int) {
-            unreachable!();
-        }
-        let sgr_params: *const u16 = (dav1d_sgr_params[lr.sgr_idx as usize]).as_ptr();
-        params.sgr.s0 = *sgr_params.offset(0) as u32;
-        params.sgr.s1 = *sgr_params.offset(1) as u32;
+        assert_eq!(lr.r#type, RAV1D_RESTORATION_SGRPROJ);
+        let sgr_params: &[u16] = &dav1d_sgr_params[lr.sgr_idx as usize];
+        params.sgr.s0 = sgr_params[0] as u32;
+        params.sgr.s1 = sgr_params[1] as u32;
         params.sgr.w0 = lr.sgr_weights[0] as i16;
-        params.sgr.w1 =
-            (128 as c_int - (lr.sgr_weights[0] as c_int + lr.sgr_weights[1] as c_int)) as i16;
-        lr_fn = (*dsp).lr.sgr[((*sgr_params.offset(0) != 0) as c_int
-            + (*sgr_params.offset(1) != 0) as c_int * 2
-            - 1) as usize];
+        params.sgr.w1 = 128 - (lr.sgr_weights[0] as i16 + lr.sgr_weights[1] as i16);
+        lr_fn = dsp.lr.sgr[(sgr_params[0] != 0) as usize + (sgr_params[1] != 0) as usize * 2 - 1];
     }
     while y + stripe_h <= row_h {
         // Change the HAVE_BOTTOM bit in edges to (sby + 1 != f->sbh || y + stripe_h != row_h)
-        edges = ::core::mem::transmute::<c_uint, LrEdgeFlags>(
-            edges as c_uint
-                ^ (-((sby + 1 != (*f).sbh || y + stripe_h != row_h) as c_int) as c_uint
-                    ^ edges as c_uint)
-                    & LR_HAVE_BOTTOM as c_int as c_uint,
-        );
+        edges = edges
+            ^ (-((sby + 1 != (*f).sbh || y + stripe_h != row_h) as c_int) as LrEdgeFlags ^ edges)
+                & LR_HAVE_BOTTOM;
         lr_fn(
-            p.cast(),
+            p.as_mut_ptr().cast(),
             stride,
-            left.cast(),
-            lpf.cast(),
+            left.as_ptr().cast(),
+            lpf.as_ptr().offset(lpf_ptr_offset).cast(),
             unit_w,
             stripe_h,
             &mut params,
             edges,
             (*f).bitdepth_max,
         );
-        left = left.offset(stripe_h as isize);
+        left = &left[stripe_h as usize..];
         y += stripe_h;
-        p = p.offset(stripe_h as isize * BD::pxstride(stride as usize) as isize);
-        edges = ::core::mem::transmute::<c_uint, LrEdgeFlags>(
-            edges as c_uint | LR_HAVE_TOP as c_int as c_uint,
-        );
+        edges = edges | LR_HAVE_TOP;
+        let p_offset = stripe_h as isize * BD::pxstride(stride as usize) as isize;
         stripe_h = cmp::min(64 >> ss_ver, row_h - y);
         if stripe_h == 0 {
             break;
         }
-        lpf = lpf.offset(4 * BD::pxstride(stride as usize) as isize);
+        p = &mut p[p_offset as usize..];
+
+        lpf = &lpf[(4 * BD::pxstride(stride as usize) as isize + lpf_ptr_offset) as usize..];
+        lpf_ptr_offset = 0;
```

Comment:
When `stride < 0` and the second iteration of this loop is reached, so `lpf_ptr_offset = 0`, then won't the slice offset be negative?

Reply:
The case `stride < 0` isn't dealt with yet

Reply:
I made changes to support the `stride < 0` case. Turns out the code is a bit simpler.

Reply:
Oh, nice!